From: Shimazu <vanessimbolon2020@gmail.com>
Date: Fri, 31 Jul 2026 02:19:04 +0000
Subject: EnchantmentTableClick: Replace SchedulerUtils.runTask with native Folia entity scheduler

AE PATCH REASON: SchedulerUtils.runTask crashes on Folia region threads (stripped class throws UnsupportedOperationException); obfuscated var1/var21 names reduce maintainability.
AE PATCH FIX: Replace SchedulerUtils.runTask with event.getEnchanter().getScheduler().run(Core.getInstance(), task -> ..., null) so book creation and view update run on the enchanter's region thread. Rename var1/var21 to event/bookItem.
AE PATCH RISK: LOW - MIGRATE: SchedulerUtils → EntityScheduler. RF: var rename. No behavior change.

---
diff --git a/net/advancedplugins/ae/enchantmentTable/EnchantmentTableClick.java b/net/advancedplugins/ae/enchantmentTable/EnchantmentTableClick.java
--- a/net/advancedplugins/ae/enchantmentTable/EnchantmentTableClick.java	2026-07-30 10:27:31.650349961 +0000
+++ b/net/advancedplugins/ae/enchantmentTable/EnchantmentTableClick.java	2026-07-30 12:23:38.769181824 +0000
@@ -15,8 +15,8 @@
 import net.advancedplugins.ae.enchanthandler.enchantments.books.BookMaker;
 import net.advancedplugins.ae.enchanthandler.enchantments.books.RandomChances;
 import net.advancedplugins.ae.impl.utils.ASManager;
-import net.advancedplugins.ae.impl.utils.SchedulerUtils;
 import net.advancedplugins.ae.impl.utils.nbt.utils.MinecraftVersion;
+import io.papermc.paper.threadedregions.scheduler.ScheduledTask;
 import net.advancedplugins.ae.utils.AManager;
 import net.advancedplugins.ae.utils.nbt.NBTapi;
 import org.bukkit.Bukkit;
@@ -35,44 +35,42 @@
       priority = EventPriority.HIGHEST,
       ignoreCancelled = true
    )
-   public void onEnchant(EnchantItemEvent var1) {
+   public void onEnchant(EnchantItemEvent event) {
       if (ETableHandler.getTableHandler().isEnabled()) {
-         if (var1.getItem().getType().name().contains("BOOK")) {
+         if (event.getItem().getType().name().contains("BOOK")) {
             if (ASManager.doChancesPass(ETableHandler.getTableHandler().getCustomEnchantBookChance())) {
-               AdvancedGroup var12 = ETableHandler.getTableHandler().getGroupForBook(var1.whichButton());
+               AdvancedGroup var12 = ETableHandler.getTableHandler().getGroupForBook(event.whichButton());
                if (var12 != null) {
                   LinkedList var13 = AEnchants.getGroupList(var12);
                   if (!var13.isEmpty()) {
                      int[] var14 = RandomChances.generate();
                      AdvancedEnchantment var16 = (AdvancedEnchantment)var13.get(ThreadLocalRandom.current().nextInt(var13.size()));
                      APloppableEnchantment var18 = new APloppableEnchantment(
-                        var14[1], var14[0], var16, ThreadLocalRandom.current().nextInt(var16.getLevels()) + 1, (Player)var1.getViewers().get(0)
-                     );
-                     ItemStack var21 = BookMaker.createABook(var18);
-                     SchedulerUtils.runTask(
-                        () -> {
-                           AdvancedEnchantItemEvent var3x = new AdvancedEnchantItemEvent(
-                              var1.getEnchanter(), var21, new HashMap<>(Map.of(var18.getEnchant(), var18.getLevel()))
-                           );
-                           Bukkit.getPluginManager().callEvent(var3x);
-                           var1.getView().setItem(0, var21);
-                        }
+                        var14[1], var14[0], var16, ThreadLocalRandom.current().nextInt(var16.getLevels()) + 1, (Player)event.getViewers().get(0)
                      );
+                     ItemStack bookItem = BookMaker.createABook(var18);
+                     event.getEnchanter().getScheduler().run(Core.getInstance(), task -> {
+                        AdvancedEnchantItemEvent enchEvent = new AdvancedEnchantItemEvent(
+                           event.getEnchanter(), bookItem, new HashMap<>(Map.of(var18.getEnchant(), var18.getLevel()))
+                        );
+                        Bukkit.getPluginManager().callEvent(enchEvent);
+                        event.getView().setItem(0, bookItem);
+                     }, null);
                   }
                }
             }
          } else {
             int var2 = ETableHandler.getTableHandler().getCustomEnchantChance();
             if (ThreadLocalRandom.current().nextInt(100) + 1 <= var2) {
-               int var3 = var1.getExpLevelCost();
+               int var3 = event.getExpLevelCost();
                int var4 = 0;
                if (MinecraftVersion.isNew()) {
-                  if (!var1.getEnchanter().hasMetadata("enchantPriceMap")) {
+                  if (!event.getEnchanter().hasMetadata("enchantPriceMap")) {
                      return;
                   }
 
-                  String var5 = ((MetadataValue)var1.getEnchanter().getMetadata("enchantPriceMap").get(0)).asString();
-                  var1.getEnchanter().removeMetadata("enchantPriceMap", Core.getInstance());
+                  String var5 = ((MetadataValue)event.getEnchanter().getMetadata("enchantPriceMap").get(0)).asString();
+                  event.getEnchanter().removeMetadata("enchantPriceMap", Core.getInstance());
                   String[] var6 = var5.split(";");
 
                   for (int var7 = 0; var7 < var6.length; var7++) {
@@ -81,13 +79,13 @@
                      }
                   }
                } else {
-                  var4 = var1.getExpLevelCost() > 16 ? 3 : (var1.getExpLevelCost() < 9 ? 1 : 2);
+                  var4 = event.getExpLevelCost() > 16 ? 3 : (event.getExpLevelCost() < 9 ? 1 : 2);
                }
 
-               ItemStack var15 = var1.getItem();
+               ItemStack var15 = event.getItem();
                ItemMeta var17 = var15.getItemMeta();
 
-               for (Entry var8 : var1.getEnchantsToAdd().entrySet()) {
+               for (Entry var8 : event.getEnchantsToAdd().entrySet()) {
                   if (!var17.hasEnchant((Enchantment)var8.getKey())) {
                      var17.addEnchant((Enchantment)var8.getKey(), (Integer)var8.getValue(), true);
                   }
@@ -108,10 +106,10 @@
                   var20--;
                }
 
-               AdvancedEnchantItemEvent var23 = new AdvancedEnchantItemEvent(var1.getEnchanter(), var15, var9);
+               AdvancedEnchantItemEvent var23 = new AdvancedEnchantItemEvent(event.getEnchanter(), var15, var9);
                Bukkit.getPluginManager().callEvent(var23);
                var9.clear();
-               var1.getView().setItem(0, var15);
+               event.getView().setItem(0, var15);
             }
          }
       }
